aten/src/ATen/core/type.cpp

```diff
@@ -909,22 +924,27 @@ bool UnionType::operator==(const Type& rhs) const {
     // Check that all the types in `this->types_` are also in
     // `union_rhs->types_`
     return std::all_of(this->types_.begin(), this->types_.end(),
-                       [&](TypePtr t) {
-                         return union_rhs->canHoldType(t);
+                       [&](TypePtr lhs_type) {
+                         return std::any_of(union_rhs->types_.begin(),
+                                            union_rhs->types_.end(),
+                                            [&](TypePtr rhs_type) {
+                                              return *lhs_type == *rhs_type;
+                                            });
                        });
   } else if (auto optional_rhs = rhs.cast<OptionalType>()) {
     if (optional_rhs->getElementType() == NumberType::get()) {
       return this->types_.size() == 4
-             && this->canHoldType(NoneType::get())
-             && this->canHoldType(NumberType::get());
+             && this->can_hold_none_
+             && can_hold_numbertype();
```

Comment:
TODO:
Maybe this can be made more generic by calling `can_hold_numbertype` inside `canHoldType`